src/blob_storage/postgres/schema.clj

```diff
@@ -1,21 +1,106 @@
 (ns blob-storage.postgres.schema
   (:refer-clojure :exclude [distinct group-by update])
   (:require [clojure.java.jdbc :as j]
-            [sqlingvo.db :as sqdb])
-  (:use sqlingvo.core))
-
+            [sqlingvo.db :as sqdb]
+            [pallet.thread-expr :as th]
+            [blob-storage.api :refer [BlobBinary]]
+            [blob-storage.coerce :refer [input-stream->byte-array]])
+  (:use sqlingvo.core)
+  (:import (org.postgresql PGConnection)
+           (org.postgresql.largeobject LargeObjectManager)
+           (java.io File FileInputStream FileOutputStream ByteArrayInputStream)))
 
 (def sqdb (sqdb/postgresql))
 
+(deftype Blob [bytes file]
+  BlobBinary
+
+  (bytea? [_]
+    (some? bytes))
+
+  (get-bytes [_]
+    (if (some? bytes)
+      bytes
+      (let [result (byte-array (.length file))]
+        (with-open [stream (FileInputStream. ^File file)]
+          (.read stream result))
+        result)))
+
+  (get-stream [_]
+    (if (some? bytes)
+      (ByteArrayInputStream. bytes)
+      (FileInputStream. ^File file))))
+
+(def large-object-threshold
+  "Value is set via blob-storage.large-object-threshold system property; defaults to
+  50 MB; uses a delay so property can be set from code after namespace is loaded but
+  before any use of the storage functions."
+  (delay (or (when-let [prop (System/getProperty "blob-storage.large-object-threshold")]
+               (Long/parseLong prop))
+             (* 50 1024 1024))))
+
+(defn- copy-streams!
+  "Like clojure.java.io/copy but works with org.postgresql.largeobject.LargeObject."
+  [input output]
+  (let [buffer (byte-array 1024)]
+    (loop []
+      (let [size (.read input buffer 0 1024)]
+        (when (pos? size)
+          (do (.write output buffer 0 size)
+              (recur)))))))
```

Comment:
esta no amerita algun test comparativo? se puede? digo para estar seguro que no hay ningun of-by-one error

Reply:
en realidad es el la misma implementación que tiene `clojure.java.io`:

```clj
(defmethod do-copy [Reader Writer] [^Reader input ^Writer output opts]
  (let [^"[C" buffer (make-array Character/TYPE (buffer-size opts))]
    (loop []
      (let [size (.read input buffer)]
        (when (pos? size)
          (do (.write output buffer 0 size)
              (recur)))))))
```

el problema es que no hay implementación para LargeObject asi que lo que hice fue robarme el código y listo.

el unico detalle es que tengo que pasarle el tamaño del buffer porque LargeObject no tiene las mismas sobrecargas que InputStream.

sobre el off-by-one, no veo razón para pensar que podría pasar con el nivel de abstracción que manejamos

Reply:
ok